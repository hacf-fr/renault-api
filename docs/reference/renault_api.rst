Renault high-level API Reference
================================

.. contents::
    :local:
    :backlinks: none


Session
---------------

.. automodule:: renault_api.renault_session
   :members:

Client
--------------

.. automodule:: renault_api.renault_client
   :members:

Account
---------------

.. automodule:: renault_api.renault_account
   :members:

Vehicle
---------------

.. automodule:: renault_api.renault_vehicle
   :members:
