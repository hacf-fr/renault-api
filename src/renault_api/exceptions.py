"""Exceptions for Renault API."""


class RenaultException(Exception):
    """Base exception for Renault API errors."""

    pass


class NotAuthenticatedException(RenaultException):
    """You are not authenticated, or authentication has expired."""

    pass
