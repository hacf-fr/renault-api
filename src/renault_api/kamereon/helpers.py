"""Helpers for Kamereon models."""
from __future__ import annotations

from typing import Any
from typing import Dict
from typing import Optional

from . import models

DAYS_OF_WEEK = [
    "monday",
    "tuesday",
    "wednesday",
    "thursday",
    "friday",
    "saturday",
    "sunday",
]


def update_schedule(schedule: models.ChargeSchedule, settings: Dict[str, Any]) -> None:
    """Update schedule."""
    if "activated" in settings:
        schedule.activated = settings["activated"]
    for day in DAYS_OF_WEEK:
        if day in settings:
            day_settings = settings[day]

            if day_settings:  # pragma: no branch
                start_time = day_settings["startTime"]
                duration = day_settings["duration"]

                setattr(
                    schedule,
                    day,
                    models.ChargeDaySchedule(day_settings, start_time, duration),
                )


def create_schedule(
    settings: Dict[str, Any]
) -> models.ChargeSchedule:  # pragma: no cover
    """Update schedule."""
    raise NotImplementedError


def get_end_time(start_time: str, duration: Optional[int] = None) -> str:
    """Compute end time."""
    total_minutes = get_total_minutes(start_time, duration)
    return format_time(total_minutes)


def format_time(total_minutes: int) -> str:
    """Format time."""
    end_hours, end_minutes = divmod(total_minutes, 60)
    end_hours = end_hours % 24
    return f"T{end_hours:02g}:{end_minutes:02g}Z"


def get_total_minutes(start_time: Optional[str], duration: Optional[int] = None) -> int:
    """Get total minutes from a `Thh:mmZ` formatted time."""
    if not start_time:  # pragma: no cover
        return 0
    return int(start_time[1:3]) * 60 + int(start_time[4:6]) + (duration or 0)
