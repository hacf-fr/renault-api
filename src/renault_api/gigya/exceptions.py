"""Gigya exceptions."""

from typing import Optional

from renault_api.exceptions import RenaultException


class GigyaException(RenaultException):
    """Base exception for Gigya errors."""

    pass


class GigyaResponseException(GigyaException):
    """Gigya returned a parsable errors."""

    def __init__(self, error_code: int, error_details: Optional[str]):
        """Initialise GigyaResponseException."""
        self.error_code = error_code
        self.error_details = error_details


class InvalidCredentialsException(GigyaResponseException):
    """Invalid loginID or password."""

    pass
