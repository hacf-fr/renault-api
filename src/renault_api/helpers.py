"""Helpers for Renault API."""
import asyncio
import functools
import logging
from typing import Dict
from typing import Optional

import aiohttp

from .const import AVAILABLE_LOCALES
from .const import CONF_GIGYA_APIKEY
from .const import CONF_GIGYA_URL
from .const import CONF_KAMEREON_APIKEY
from .const import CONF_KAMEREON_URL
from .const import LOCALE_BASE_URL
from .exceptions import RenaultException

_LOGGER = logging.getLogger(__package__)


async def get_api_keys(
    locale: str,
    force_load: bool = False,
    websession: Optional[aiohttp.ClientSession] = None,
) -> Dict[str, str]:
    """Get the API keys for specified locale.

    Args:
        locale (str): locale code (preferrably from AVAILABLE_LOCALES.keys())
        force_load (bool): bypass internal AVAILABLE_LOCALES
        websession (aiohttp.ClientSession): required if locale not in AVAILABLE_LOCALES

    Returns:
        Dict with gigya-api-key, gigya-api-url,
        kamereon-api-key and kamereon-api-url

    Raises:
        RenaultException: an issue occured loading the API keys
    """
    if locale in AVAILABLE_LOCALES.keys() and not force_load:
        return AVAILABLE_LOCALES[locale]
    else:
        _LOGGER.warning(
            "Locale %s was not found in AVAILABLE_LOCALES "
            "(or force_load used). "
            "Attempting to load details from Renault servers.",
            locale,
        )
        if websession is None:
            raise RenaultException("aiohttp_session is not set.")

        url = f"{LOCALE_BASE_URL}/configuration/android/config_{locale}.json"
        async with websession.get(url) as response:  # pragma: no cover
            try:
                response.raise_for_status()
            except aiohttp.ClientResponseError as exc:
                raise RenaultException(
                    f"Locale not found on Renault server (HTTPStatus = {exc.status})."
                ) from exc
            # Server sometimes returns invalid content-type
            # eg. application/octet-stream
            response_body = await response.json(content_type=None)

            _LOGGER.debug(
                "Received api keys from myrenault response: %s", response_body
            )

            servers = response_body["servers"]
            return {
                CONF_GIGYA_APIKEY: servers["gigyaProd"]["apikey"],
                CONF_GIGYA_URL: servers["gigyaProd"]["target"],
                CONF_KAMEREON_APIKEY: servers["wiredProd"]["apikey"],
                CONF_KAMEREON_URL: servers["wiredProd"]["target"],
            }


def create_aiohttp_closed_event(
    websession: aiohttp.ClientSession,
) -> asyncio.Event:  # pragma: no cover
    """Work around aiohttp issue that doesn't properly close transports on exit.

    See https://github.com/aio-libs/aiohttp/issues/1925#issuecomment-639080209

    Args:
        websession (aiohttp.ClientSession): session for which to generate the event.

    Returns:
        An event that will be set once all transports have been properly closed.
    """
    transports = 0
    all_is_lost = asyncio.Event()

    def connection_lost(exc, orig_lost):  # type: ignore
        nonlocal transports

        try:
            orig_lost(exc)
        finally:
            transports -= 1
            if transports == 0:
                all_is_lost.set()

    def eof_received(orig_eof_received):  # type: ignore
        try:
            orig_eof_received()
        except AttributeError:
            # It may happen that eof_received() is called after
            # _app_protocol and _transport are set to None.
            pass

    for conn in websession.connector._conns.values():  # type: ignore
        for handler, _ in conn:
            proto = getattr(handler.transport, "_ssl_protocol", None)
            if proto is None:
                continue

            transports += 1
            orig_lost = proto.connection_lost
            orig_eof_received = proto.eof_received

            proto.connection_lost = functools.partial(
                connection_lost, orig_lost=orig_lost
            )
            proto.eof_received = functools.partial(
                eof_received, orig_eof_received=orig_eof_received
            )

    if transports == 0:
        all_is_lost.set()

    return all_is_lost
