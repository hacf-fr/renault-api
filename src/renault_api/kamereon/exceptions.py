"""Kamereon exceptions."""
from typing import Optional

from renault_api.exceptions import RenaultException


class KamereonException(RenaultException):
    """Base exception for Kamereon errors."""

    pass


class KamereonResponseException(KamereonException):
    """Kamereon returned a parsable errors."""

    def __init__(self, error_code: Optional[str], error_details: Optional[str]):
        """Initialise KamereonResponseException."""
        self.error_code = error_code
        self.error_details = error_details


class AccessDeniedException(KamereonResponseException):
    """Access is denied for this resource."""

    pass


class NotSupportedException(KamereonResponseException):
    """This feature is not technically supported by this gateway."""

    pass


class InvalidUpstreamException(KamereonResponseException):
    """Invalid response from the upstream server."""

    pass


class QuotaLimitException(KamereonResponseException):
    """You have reached your quota limit."""

    pass


class InvalidInputException(KamereonResponseException):
    """The input is invalid."""

    pass


class ResourceNotFoundException(KamereonResponseException):
    """Resource not found."""

    pass


class FailedForwardException(KamereonResponseException):
    """Failed to forward request to remote service."""

    pass
