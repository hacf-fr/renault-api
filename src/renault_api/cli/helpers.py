"""Helpers for Renault API."""
import asyncio
import functools
from datetime import datetime
from datetime import timedelta
from typing import Any
from typing import Callable
from typing import Optional
from typing import Tuple

import aiohttp
import click
import dateparser
import tzlocal

from renault_api.exceptions import RenaultException
from renault_api.kamereon.helpers import DAYS_OF_WEEK


_DATETIME_FORMAT = "%Y-%m-%d %H:%M:%S"


def coro_with_websession(func: Callable[..., Any]) -> Callable[..., Any]:
    """Ensure the routine runs on an event loop with a websession."""

    async def run_command(func: Callable[..., Any], *args: Any, **kwargs: Any) -> None:
        async with aiohttp.ClientSession() as websession:
            try:
                kwargs["websession"] = websession
                await func(*args, **kwargs)
            except RenaultException as exc:  # pragma: no cover
                raise click.ClickException(str(exc)) from exc
            finally:
                closed_event = create_aiohttp_closed_event(websession)
                await websession.close()
                await closed_event.wait()

    def wrapper(*args: Any, **kwargs: Any) -> None:
        asyncio.run(run_command(func, *args, **kwargs))

    return functools.update_wrapper(wrapper, func)


def days_of_week_option(helptext: str) -> Callable[..., Any]:
    """Add day of week string options."""

    def decorator(func: Callable[..., Any]) -> Callable[..., Any]:
        for day in reversed(DAYS_OF_WEEK):
            func = click.option(
                f"--{day}",
                help=helptext.format(day.capitalize()),
            )(func)
        return func

    return decorator


def start_end_option(add_period: bool) -> Callable[..., Any]:
    """Add start/end options."""

    def decorator(func: Callable[..., Any]) -> Callable[..., Any]:
        func = click.option(
            "--from", "start", help="Date to start showing history from", required=True
        )(func)
        func = click.option(
            "--to",
            "end",
            help="Date to finish showing history at (cannot be in the future)",
            required=True,
        )(func)
        if add_period:
            func = click.option(
                "--period",
                default="month",
                help="Period over which to aggregate.",
                type=click.Choice(["day", "month"], case_sensitive=False),
            )(func)
        return func

    return decorator


def create_aiohttp_closed_event(
    websession: aiohttp.ClientSession,
) -> asyncio.Event:  # pragma: no cover
    """Work around aiohttp issue that doesn't properly close transports on exit.

    See https://github.com/aio-libs/aiohttp/issues/1925#issuecomment-639080209

    Args:
        websession (aiohttp.ClientSession): session for which to generate the event.

    Returns:
        An event that will be set once all transports have been properly closed.
    """
    transports = 0
    all_is_lost = asyncio.Event()

    def connection_lost(exc, orig_lost):  # type: ignore
        nonlocal transports

        try:
            orig_lost(exc)
        finally:
            transports -= 1
            if transports == 0:
                all_is_lost.set()

    def eof_received(orig_eof_received):  # type: ignore
        try:
            orig_eof_received()
        except AttributeError:
            # It may happen that eof_received() is called after
            # _app_protocol and _transport are set to None.
            pass

    for conn in websession.connector._conns.values():  # type: ignore
        for handler, _ in conn:
            proto = getattr(handler.transport, "_ssl_protocol", None)
            if proto is None:
                continue

            transports += 1
            orig_lost = proto.connection_lost
            orig_eof_received = proto.eof_received

            proto.connection_lost = functools.partial(
                connection_lost, orig_lost=orig_lost
            )
            proto.eof_received = functools.partial(
                eof_received, orig_eof_received=orig_eof_received
            )

    if transports == 0:
        all_is_lost.set()

    return all_is_lost


def parse_dates(start: str, end: str) -> Tuple[datetime, datetime]:
    """Convert start/end string arguments into datetime arguments."""
    parsed_start = dateparser.parse(start)
    parsed_end = dateparser.parse(end)

    if not parsed_start:  # pragma: no cover
        raise ValueError(f"Unable to parse `{start}` into start datetime.")
    if not parsed_end:  # pragma: no cover
        raise ValueError(f"Unable to parse `{end}` into end datetime.")

    return (parsed_start, parsed_end)


def _timezone_offset() -> int:
    """Return UTC offset in minutes."""
    utcoffset = tzlocal.get_localzone().utcoffset(datetime.now())
    if utcoffset:
        return int(utcoffset.total_seconds() / 60)
    return 0  # pragma: no cover


def _format_tzdatetime(date_string: str) -> str:
    date = datetime.fromisoformat(date_string.replace("Z", "+00:00"))
    return str(date.astimezone(tzlocal.get_localzone()).strftime(_DATETIME_FORMAT))


def _format_tztime(time: str) -> str:
    total_minutes = int(time[1:3]) * 60 + int(time[4:6]) + _timezone_offset()
    hours, minutes = divmod(total_minutes, 60)
    hours = hours % 24  # Ensure it is 00-23
    return f"{hours:02g}:{minutes:02g}"


def convert_minutes_to_tztime(minutes: int) -> str:
    """Convert minutes to Thh:mmZ format."""
    total_minutes = minutes - _timezone_offset()
    hours, minutes = divmod(total_minutes, 60)
    hours = hours % 24  # Ensure it is 00-23
    return f"T{hours:02g}:{minutes:02g}Z"


def _format_minutes(mins: float) -> str:
    d = timedelta(minutes=mins)
    return str(d)


def get_display_value(
    value: Optional[Any] = None,
    unit: Optional[str] = None,
) -> str:
    """Get a display for value."""
    if value is None:  # pragma: no cover
        return ""
    if unit is None:
        return str(value)
    if unit == "tzdatetime":
        return _format_tzdatetime(value)
    if unit == "tztime":
        return _format_tztime(value)
    if unit == "minutes":
        return _format_minutes(value)
    if unit == "kW":
        value = value / 1000
        return f"{value:.2f} {unit}"
    return f"{value} {unit}"
