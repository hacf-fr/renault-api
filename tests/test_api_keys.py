"""Test cases for the Renault client API keys."""
import pytest
from aiohttp import ClientSession
from aioresponses import aioresponses

from renault_api.const import AVAILABLE_LOCALES
from renault_api.const import CONF_GIGYA_APIKEY
from renault_api.const import CONF_GIGYA_URL
from renault_api.const import CONF_KAMEREON_APIKEY
from renault_api.const import CONF_KAMEREON_URL
from renault_api.const import LOCALE_BASE_URL
from renault_api.exceptions import RenaultException
from renault_api.helpers import get_api_keys


@pytest.mark.asyncio
@pytest.mark.parametrize("locale", AVAILABLE_LOCALES.keys())
async def test_available_locales(locale: str) -> None:
    """Ensure all items AVAILABLE_LOCALES have correct data."""
    expected_api_keys = AVAILABLE_LOCALES[locale]

    api_keys = await get_api_keys(locale)
    assert api_keys == expected_api_keys
    for key in [
        CONF_GIGYA_APIKEY,
        CONF_GIGYA_URL,
        CONF_KAMEREON_APIKEY,
        CONF_KAMEREON_URL,
    ]:
        assert api_keys[key]


@pytest.mark.asyncio
async def test_missing_aiohttp_session() -> None:
    """Ensure failure to unknown locale if aiohttp_session is not set."""
    locale = "invalid"

    with pytest.raises(RenaultException) as excinfo:
        await get_api_keys(locale)
    assert "aiohttp_session is not set." in str(excinfo)


@pytest.mark.asyncio
@pytest.mark.parametrize("locale", AVAILABLE_LOCALES.keys())
@pytest.mark.skip(reason="Makes real calls to Renault servers")
async def test_preload_force_api_keys(websession: ClientSession, locale: str) -> None:
    """Ensure is able to parse a valid locale from Renault servers."""
    expected_api_keys = AVAILABLE_LOCALES[locale]

    api_keys = await get_api_keys(locale, True, websession)

    assert api_keys == expected_api_keys


@pytest.mark.asyncio
async def test_preload_unknown_api_keys(websession: ClientSession) -> None:
    """Ensure is able to parse a known known."""
    expected_api_keys = AVAILABLE_LOCALES["fr_FR"]

    fake_locale = "invalid"
    fake_url = f"{LOCALE_BASE_URL}/configuration/android/config_{fake_locale}.json"
    with open("tests/fixtures/config_sample.txt", "r") as f:
        fake_body = f.read()

    with aioresponses() as mocked_responses:
        mocked_responses.get(fake_url, status=200, body=fake_body)

        api_keys = await get_api_keys(fake_locale, aiohttp_session=websession)

        assert api_keys == expected_api_keys


@pytest.mark.asyncio
async def test_preload_invalid_api_keys(websession: ClientSession) -> None:
    """Ensure is able to parse an invalid locale."""
    fake_locale = "fake"
    fake_url = f"{LOCALE_BASE_URL}/configuration/android/config_{fake_locale}.json"

    with aioresponses() as mocked_responses:
        mocked_responses.get(fake_url, status=404)

        with pytest.raises(RenaultException) as excinfo:
            await get_api_keys(fake_locale, aiohttp_session=websession)
        assert "Locale not found on Renault server" in str(excinfo)
