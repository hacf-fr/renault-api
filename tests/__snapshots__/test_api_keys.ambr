# serializer version: 1
# name: test_available_locales[bg_BG]
  dict({
    'gigya-api-key': '3__3ER_6lFvXEXHTP_faLtq6eEdbKDXd9F5GoKwzRyZq37ZQ-db7mXcLzR1Jtls5sn',
    'gigya-root-url': 'https://accounts.eu1.gigya.com',
    'kamereon-api-key': 'YjkKtHmGfaceeuExUDKGxrLZGGvtVS0J',
    'kamereon-root-url': 'https://api-wired-prod-1-euw1.wrd-aws.com',
  })
# ---
# name: test_available_locales[cs_CZ]
  dict({
    'gigya-api-key': '3_oRlKr5PCVL_sPWUZdJ8c5NOl5Ej8nIZw7VKG7S9Rg36UkDszFzfHfxCaUAUU5or2',
    'gigya-root-url': 'https://accounts.eu1.gigya.com',
    'kamereon-api-key': 'YjkKtHmGfaceeuExUDKGxrLZGGvtVS0J',
    'kamereon-root-url': 'https://api-wired-prod-1-euw1.wrd-aws.com',
  })
# ---
# name: test_available_locales[da_DK]
  dict({
    'gigya-api-key': '3_5x-2C8b1R4MJPQXkwTPdIqgBpcw653Dakw_ZaEneQRkTBdg9UW9Qg_5G-tMNrTMc',
    'gigya-root-url': 'https://accounts.eu1.gigya.com',
    'kamereon-api-key': 'YjkKtHmGfaceeuExUDKGxrLZGGvtVS0J',
    'kamereon-root-url': 'https://api-wired-prod-1-euw1.wrd-aws.com',
  })
# ---
# name: test_available_locales[de_AT]
  dict({
    'gigya-api-key': '3__B4KghyeUb0GlpU62ZXKrjSfb7CPzwBS368wioftJUL5qXE0Z_sSy0rX69klXuHy',
    'gigya-root-url': 'https://accounts.eu1.gigya.com',
    'kamereon-api-key': 'YjkKtHmGfaceeuExUDKGxrLZGGvtVS0J',
    'kamereon-root-url': 'https://api-wired-prod-1-euw1.wrd-aws.com',
  })
# ---
# name: test_available_locales[de_CH]
  dict({
    'gigya-api-key': '3_UyiWZs_1UXYCUqK_1n7l7l44UiI_9N9hqwtREV0-UYA_5X7tOV-VKvnGxPBww4q2',
    'gigya-root-url': 'https://accounts.eu1.gigya.com',
    'kamereon-api-key': 'YjkKtHmGfaceeuExUDKGxrLZGGvtVS0J',
    'kamereon-root-url': 'https://api-wired-prod-1-euw1.wrd-aws.com',
  })
# ---
# name: test_available_locales[de_DE]
  dict({
    'gigya-api-key': '3_7PLksOyBRkHv126x5WhHb-5pqC1qFR8pQjxSeLB6nhAnPERTUlwnYoznHSxwX668',
    'gigya-root-url': 'https://accounts.eu1.gigya.com',
    'kamereon-api-key': 'YjkKtHmGfaceeuExUDKGxrLZGGvtVS0J',
    'kamereon-root-url': 'https://api-wired-prod-1-euw1.wrd-aws.com',
  })
# ---
# name: test_available_locales[en_GB]
  dict({
    'gigya-api-key': '3_e8d4g4SE_Fo8ahyHwwP7ohLGZ79HKNN2T8NjQqoNnk6Epj6ilyYwKdHUyCw3wuxz',
    'gigya-root-url': 'https://accounts.eu1.gigya.com',
    'kamereon-api-key': 'YjkKtHmGfaceeuExUDKGxrLZGGvtVS0J',
    'kamereon-root-url': 'https://api-wired-prod-1-euw1.wrd-aws.com',
  })
# ---
# name: test_available_locales[en_IE]
  dict({
    'gigya-api-key': '3_Xn7tuOnT9raLEXuwSI1_sFFZNEJhSD0lv3gxkwFtGI-RY4AgiePBiJ9EODh8d9yo',
    'gigya-root-url': 'https://accounts.eu1.gigya.com',
    'kamereon-api-key': 'YjkKtHmGfaceeuExUDKGxrLZGGvtVS0J',
    'kamereon-root-url': 'https://api-wired-prod-1-euw1.wrd-aws.com',
  })
# ---
# name: test_available_locales[es_ES]
  dict({
    'gigya-api-key': '3_DyMiOwEaxLcPdBTu63Gv3hlhvLaLbW3ufvjHLeuU8U5bx3zx19t5rEKq7KMwk9f1',
    'gigya-root-url': 'https://accounts.eu1.gigya.com',
    'kamereon-api-key': 'YjkKtHmGfaceeuExUDKGxrLZGGvtVS0J',
    'kamereon-root-url': 'https://api-wired-prod-1-euw1.wrd-aws.com',
  })
# ---
# name: test_available_locales[es_MX]
  dict({
    'gigya-api-key': '3_BFzR-2wfhMhUs5OCy3R8U8IiQcHS-81vF8bteSe8eFrboMTjEWzbf4pY1aHQ7cW0',
    'gigya-root-url': 'https://accounts.us1.gigya.com',
    'kamereon-api-key': 'YjkKtHmGfaceeuExUDKGxrLZGGvtVS0J',
    'kamereon-root-url': 'https://api-wired-prod-1-usw2.wrd-aws.com',
  })
# ---
# name: test_available_locales[fi_FI]
  dict({
    'gigya-api-key': '3_xSRCLDYhk1SwSeYQLI3DmA8t-etfAfu5un51fws125ANOBZHgh8Lcc4ReWSwaqNY',
    'gigya-root-url': 'https://accounts.eu1.gigya.com',
    'kamereon-api-key': 'YjkKtHmGfaceeuExUDKGxrLZGGvtVS0J',
    'kamereon-root-url': 'https://api-wired-prod-1-euw1.wrd-aws.com',
  })
# ---
# name: test_available_locales[fr_BE]
  dict({
    'gigya-api-key': '3_ZK9x38N8pzEvdiG7ojWHeOAAej43APkeJ5Av6VbTkeoOWR4sdkRc-wyF72HzUB8X',
    'gigya-root-url': 'https://accounts.eu1.gigya.com',
    'kamereon-api-key': 'YjkKtHmGfaceeuExUDKGxrLZGGvtVS0J',
    'kamereon-root-url': 'https://api-wired-prod-1-euw1.wrd-aws.com',
  })
# ---
# name: test_available_locales[fr_CH]
  dict({
    'gigya-api-key': '3_h3LOcrKZ9mTXxMI9clb2R1VGAWPke6jMNqMw4yYLz4N7PGjYyD0hqRgIFAIHusSn',
    'gigya-root-url': 'https://accounts.eu1.gigya.com',
    'kamereon-api-key': 'YjkKtHmGfaceeuExUDKGxrLZGGvtVS0J',
    'kamereon-root-url': 'https://api-wired-prod-1-euw1.wrd-aws.com',
  })
# ---
# name: test_available_locales[fr_FR]
  dict({
    'gigya-api-key': '3_4LKbCcMMcvjDm3X89LU4z4mNKYKdl_W0oD9w-Jvih21WqgJKtFZAnb9YdUgWT9_a',
    'gigya-root-url': 'https://accounts.eu1.gigya.com',
    'kamereon-api-key': 'YjkKtHmGfaceeuExUDKGxrLZGGvtVS0J',
    'kamereon-root-url': 'https://api-wired-prod-1-euw1.wrd-aws.com',
  })
# ---
# name: test_available_locales[fr_LU]
  dict({
    'gigya-api-key': '3_zt44Wl_wT9mnqn-BHrR19PvXj3wYRPQKLcPbGWawlatFR837KdxSZZStbBTDaqnb',
    'gigya-root-url': 'https://accounts.eu1.gigya.com',
    'kamereon-api-key': 'YjkKtHmGfaceeuExUDKGxrLZGGvtVS0J',
    'kamereon-root-url': 'https://api-wired-prod-1-euw1.wrd-aws.com',
  })
# ---
# name: test_available_locales[hr_HR]
  dict({
    'gigya-api-key': '3_HcDC5GGZ89NMP1jORLhYNNCcXt7M3thhZ85eGrcQaM2pRwrgrzcIRWEYi_36cFj9',
    'gigya-root-url': 'https://accounts.eu1.gigya.com',
    'kamereon-api-key': 'YjkKtHmGfaceeuExUDKGxrLZGGvtVS0J',
    'kamereon-root-url': 'https://api-wired-prod-1-euw1.wrd-aws.com',
  })
# ---
# name: test_available_locales[hu_HU]
  dict({
    'gigya-api-key': '3_nGDWrkSGZovhnVFv5hdIxyuuCuJGZfNmlRGp7-5kEn9yb0bfIfJqoDa2opHOd3Mu',
    'gigya-root-url': 'https://accounts.eu1.gigya.com',
    'kamereon-api-key': 'YjkKtHmGfaceeuExUDKGxrLZGGvtVS0J',
    'kamereon-root-url': 'https://api-wired-prod-1-euw1.wrd-aws.com',
  })
# ---
# name: test_available_locales[it_CH]
  dict({
    'gigya-api-key': '3_gHkmHaGACxSLKXqD_uDDx415zdTw7w8HXAFyvh0qIP0WxnHPMF2B9K_nREJVSkGq',
    'gigya-root-url': 'https://accounts.eu1.gigya.com',
    'kamereon-api-key': 'YjkKtHmGfaceeuExUDKGxrLZGGvtVS0J',
    'kamereon-root-url': 'https://api-wired-prod-1-euw1.wrd-aws.com',
  })
# ---
# name: test_available_locales[it_IT]
  dict({
    'gigya-api-key': '3_js8th3jdmCWV86fKR3SXQWvXGKbHoWFv8NAgRbH7FnIBsi_XvCpN_rtLcI07uNuq',
    'gigya-root-url': 'https://accounts.eu1.gigya.com',
    'kamereon-api-key': 'YjkKtHmGfaceeuExUDKGxrLZGGvtVS0J',
    'kamereon-root-url': 'https://api-wired-prod-1-euw1.wrd-aws.com',
  })
# ---
# name: test_available_locales[nl_BE]
  dict({
    'gigya-api-key': '3_yachztWczt6i1pIMhLIH9UA6DXK6vXXuCDmcsoA4PYR0g35RvLPDbp49YribFdpC',
    'gigya-root-url': 'https://accounts.eu1.gigya.com',
    'kamereon-api-key': 'YjkKtHmGfaceeuExUDKGxrLZGGvtVS0J',
    'kamereon-root-url': 'https://api-wired-prod-1-euw1.wrd-aws.com',
  })
# ---
# name: test_available_locales[nl_NL]
  dict({
    'gigya-api-key': '3_ZIOtjqmP0zaHdEnPK7h1xPuBYgtcOyUxbsTY8Gw31Fzy7i7Ltjfm-hhPh23fpHT5',
    'gigya-root-url': 'https://accounts.eu1.gigya.com',
    'kamereon-api-key': 'YjkKtHmGfaceeuExUDKGxrLZGGvtVS0J',
    'kamereon-root-url': 'https://api-wired-prod-1-euw1.wrd-aws.com',
  })
# ---
# name: test_available_locales[no_NO]
  dict({
    'gigya-api-key': '3_QrPkEJr69l7rHkdCVls0owC80BB4CGz5xw_b0gBSNdn3pL04wzMBkcwtbeKdl1g9',
    'gigya-root-url': 'https://accounts.eu1.gigya.com',
    'kamereon-api-key': 'YjkKtHmGfaceeuExUDKGxrLZGGvtVS0J',
    'kamereon-root-url': 'https://api-wired-prod-1-euw1.wrd-aws.com',
  })
# ---
# name: test_available_locales[pl_PL]
  dict({
    'gigya-api-key': '3_2YBjydYRd1shr6bsZdrvA9z7owvSg3W5RHDYDp6AlatXw9hqx7nVoanRn8YGsBN8',
    'gigya-root-url': 'https://accounts.eu1.gigya.com',
    'kamereon-api-key': 'YjkKtHmGfaceeuExUDKGxrLZGGvtVS0J',
    'kamereon-root-url': 'https://api-wired-prod-1-euw1.wrd-aws.com',
  })
# ---
# name: test_available_locales[pt_PT]
  dict({
    'gigya-api-key': '3__afxovspi2-Ip1E5kNsAgc4_35lpLAKCF6bq4_xXj2I2bFPjIWxAOAQJlIkreKTD',
    'gigya-root-url': 'https://accounts.eu1.gigya.com',
    'kamereon-api-key': 'YjkKtHmGfaceeuExUDKGxrLZGGvtVS0J',
    'kamereon-root-url': 'https://api-wired-prod-1-euw1.wrd-aws.com',
  })
# ---
# name: test_available_locales[ro_RO]
  dict({
    'gigya-api-key': '3_WlBp06vVHuHZhiDLIehF8gchqbfegDJADPQ2MtEsrc8dWVuESf2JCITRo5I2CIxs',
    'gigya-root-url': 'https://accounts.eu1.gigya.com',
    'kamereon-api-key': 'YjkKtHmGfaceeuExUDKGxrLZGGvtVS0J',
    'kamereon-root-url': 'https://api-wired-prod-1-euw1.wrd-aws.com',
  })
# ---
# name: test_available_locales[ru_RU]
  dict({
    'gigya-api-key': '3_N_ecy4iDyoRtX8v5xOxewwZLKXBjRgrEIv85XxI0KJk8AAdYhJIi17LWb086tGXR',
    'gigya-root-url': 'https://accounts.eu1.gigya.com',
    'kamereon-api-key': 'YjkKtHmGfaceeuExUDKGxrLZGGvtVS0J',
    'kamereon-root-url': 'https://api-wired-prod-1-euw1.wrd-aws.com',
  })
# ---
# name: test_available_locales[sk_SK]
  dict({
    'gigya-api-key': '3_e8d4g4SE_Fo8ahyHwwP7ohLGZ79HKNN2T8NjQqoNnk6Epj6ilyYwKdHUyCw3wuxz',
    'gigya-root-url': 'https://accounts.eu1.gigya.com',
    'kamereon-api-key': 'YjkKtHmGfaceeuExUDKGxrLZGGvtVS0J',
    'kamereon-root-url': 'https://api-wired-prod-1-euw1.wrd-aws.com',
  })
# ---
# name: test_available_locales[sl_SI]
  dict({
    'gigya-api-key': '3_QKt0ADYxIhgcje4F3fj9oVidHsx3JIIk-GThhdyMMQi8AJR0QoHdA62YArVjbZCt',
    'gigya-root-url': 'https://accounts.eu1.gigya.com',
    'kamereon-api-key': 'YjkKtHmGfaceeuExUDKGxrLZGGvtVS0J',
    'kamereon-root-url': 'https://api-wired-prod-1-euw1.wrd-aws.com',
  })
# ---
# name: test_available_locales[sv_SE]
  dict({
    'gigya-api-key': '3_EN5Hcnwanu9_Dqot1v1Aky1YelT5QqG4TxveO0EgKFWZYu03WkeB9FKuKKIWUXIS',
    'gigya-root-url': 'https://accounts.eu1.gigya.com',
    'kamereon-api-key': 'YjkKtHmGfaceeuExUDKGxrLZGGvtVS0J',
    'kamereon-root-url': 'https://api-wired-prod-1-euw1.wrd-aws.com',
  })
# ---
